vignettes/articles/writing-custom-fns.Rmd

```diff
@@ -0,0 +1,244 @@
+
+</div>
+
+## Arguments to custom check functions
+
+### Arguments available in the caller environment 
+
+The `hubValidations` package automatically passes a set of standard objects to custom check functions. These objects are passed to arguments of custom check functions with the same name and do not need to be included in the `args` configuration in the `validations.yml` file.
+
+
+```{r child="children/_custom-fn-available-args.Rmd", echo=FALSE, results="asis"}
+```
+
+### Additional arguments
+
+You can add additional arguments to custom check functions and pass values to them by including them in the `args` configuration in the `validations.yml` file. These values are passed to the custom check function by `hubValidations` when the function is called.
+
+
+If you do add additional arguments to a custom check function, you should also add input  checks at the start of the function to ensure inputs are valid. Package `checkmate` contains a wide range of functions for checking inputs.
+
+For example, the optional check `opt_check_tbl_col_timediff()` (which is deployed in exactly the same fashion as custom functions, i.e. through the `validations.yml` file) takes additional arguments `t0_colname`, `t1_colname` and `timediff`. 
+
+
+```{r}
+opt_check_tbl_col_timediff
+```
+
+
+
+You can add an example extra argument with `extra_args = TRUE` when creating the custom check function with `create_custom_check()`.
+
+```{r}
+create_custom_check("cstm_check_tbl_args",
+  hub_path = hub_path,
+  extra_args = TRUE
+)
+```
+
+This adds an extra argument `extra_arg` to the custom check function formals as well as an example input check.
+s
+```{r, code=readLines(r_path(hub_path, "cstm_check_tbl_args")), eval=FALSE}
+```
+
+
+## Function output
+
+### Capturing and returning check results
+
+**Returning a check condition with `capture_check_cnd()`**
+
+The `capture_check_cnd()` function is used to return a check condition and it's output should also be what a custom check function returns in most cases. The function returns a `<hub_check>` class object depending on the value passed to the `check` argument, which represents the summary of the condition being checked by a given validation function. 
+
+If the value passed to `check ` is `TRUE`, the function returns a `<message/check_success>` class object. 
+
+If the value is `FALSE`, the output depends on the `error` argument.  
+
+- If `error` is `TRUE`, the function returns a `<message/check_error>` class object. If `error` is `FALSE`, the function returns a `<error/check_failure>` class object, which indicates the check has failed. 
+- If `error` is `TRUE`, the function returns a `<error/check_error>` class object, which indicates the check has failed and additionally causes execution of further custom validation functions to halt. Set `error = TRUE` if downstream checks cannot be run if the current check fails.
```

Comment:
I might open an issue for my question, I don't think here is the best place, but I thought we also had a warning option? If not, should we add one? It it might be useful to returns that information for an issue that is not technically incorrect but might be worth checking. For example, we might have a cumulative case projection bigger than the population size, it might be correct due to re-infection but it might be worth checking to be sure, so it might be good to returns a warning in that case, does that make sense?

Reply:
Thank you for the information, and I apologize I miss that one.  
I have a small issue with the warning being a failure because it technically did not fail, but it's more a philosophical question. To continue my example, if I want to be able to tag projection of cases that are higher than population size. It can be technically correct as people can be re-infected but it will be a surprising projections (in this context) so I want to be able to tag it so the team can easily check that it's what they want to submits.

Reply:
> I want to be able to tag it so the team can easily check that it's what they want to submits.

This is what `<error/check_failure>` (what used to be `<warning/check_failure>` is doing). In the old version of hubValidations, the `<warning/check_failure>` still gave an `Error` in the overall validation result. 

It mostly boils down to semantics. **Regardless of the name (warning or error), they both resulted in an overall validation that did not pass**. 

Since this requires human intervention, if this is a custom validation function, then you can add a bold comment that says "projected cases (n) higher than population size (N) **NOTE: reinfections can cause projections to be higher than the population size**, if this is the case, then the error is a false-positive". Then you as the admin can confirm the following and merge the pull request, overriding the failed check:

1. this is the only check that fails
2. you and the modeler are certain that reinfections are the cause

Reply:
thanks for the clarification! that make sense, thanks. I think I just worry that people misunderstood the error as something that need to be fixed and is incorrect. But, I guess we can put information in the resulting message to clarify that.

Reply:
@LucieContamin`warning/check_failure>`s have always been considered failures that need fixing in `hubValidations` and would cause PR workflows to fail (as per @zkamvar's comment). Really what you are describing is a completely new feature in `hubValidations` (i.e. something that is not quite a pass but is not a failure ether).

If this is something you might indeed use, I suggest you open an issue about it describing how see the feature being used. In fact, the changes made in 0.6.0 actually allow for the use of warnings for such a purpose now.